cpp/src/lance/format/schema.cc

```diff
@@ -383,6 +385,38 @@ std::shared_ptr<Field> Field::Project(const std::shared_ptr<::arrow::Field>& arr
   return new_field;
 }
 
+::arrow::Result<std::shared_ptr<Field>> Field::Merge(const ::arrow::Field& arrow_field) const {
+  if (name() != arrow_field.name()) {
+    return ::arrow::Status::Invalid(
+        "Attempt to merge two different fields: ", name(), "!=", arrow_field.name());
+  }
+  auto self_type = type();
+  if (self_type->id() != arrow_field.type()->id()) {
+    return ::arrow::Status::Invalid(
+        "Can not merge two fields: ", self_type->ToString(), "!=", arrow_field.type()->ToString());
+  };
+  auto new_field = Copy(true);
+  if (::arrow::is_list_like(self_type->id())) {
+    auto list_type = std::dynamic_pointer_cast<::arrow::ListType>(arrow_field.type());
+    return children_[0]->Merge(*list_type->value_field());
+  } else if (lance::arrow::is_struct(self_type)) {
+    auto struct_type = std::dynamic_pointer_cast<::arrow::StructType>(arrow_field.type());
+    for (auto& arrow_child : struct_type->fields()) {
+      auto child = Get(arrow_child->name());
+      if (!child) {
+        new_field->children_.emplace_back(std::make_shared<Field>(arrow_child));
+      } else {
+        ARROW_ASSIGN_OR_RAISE(auto new_child_field, child->Merge(*arrow_child));
+        new_field->children_.emplace_back(new_child_field);
+      }
```

Comment:
what about children of this struct field that's not in the arrow_field?

Reply:
`new_field` has the children copied here https://github.com/eto-ai/lance/pull/263/files/90ec4c778ad42bf0f54d1abb0b8aef09e45bdd22#diff-0459f40916915b924d3c6aa11554911c9f00fcded2c0361b85f215f6623cce58R398.

It seems to a bug that the common child between two fields will be insert twice tho. Lemme fix it.